### PR TITLE
Improve notebook editor performance

### LINKED ISSUE #33676: Performance regression in how we parse query_metadata response leads to notebook being unusable
### Describe the bug

We introduced a regression in the query_metadata, possibly because of several factors, that lead questions with lots of Metadata to slow down heavily the frontend (Pro tip: use the table metadata and generate a large lorem ipsum on the fields and tables descriptions to make a heavy response along with a "big" question with lots of stages and fields. remember we call the table_metadata for every linked table on a query, so if table 1 has a FK to table 2, we fetch both tables metadata)

Performance in 46:
image

Profiler
image
image

Now let's move to 47:
look at the separation between calls (sizes of responses stayed the same)
image

profiler:
2X
image

This will get amplified as more tables are joined in the data model

### To Reproduce

1) go to the data model and try to join a single table to all others. Also try to complete as much metadata as possible
2) then create a GUI question and add that single table, build a question with lots of aggregations, fields, etc
3) compare it between 46 and 47, see the difference in rendering

### Expected behavior

UI should be F A S T

### Logs

NA

### Information about your Metabase installation

```JSON
v47
```

### Severity

P1

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by metamben**
Fixes #33676.

**How to test**

Use @ranquild's repro from #33676 with and without these changes.

**Known problems**

Using the same query, try hiding and showing columns on the viz settings page. It will be very slow, but MLv2 is not involved there and there is no obvious bottleneck (see my [comment](https://github.com/metabase/metabase/issues/33676#issuecomment-1708779117) on the issue).

### CODE REVIEW

File: shared/src/metabase/mbql/util.cljc [hunk 1]
```diff
@@ -64,53 +64,55 @@
...
-    ;; for `not` that wraps another `not`, eliminate both
-    [:not [:not arg]]
-    (recur arg)
-
-    :else
-    filter-clause))
+  [x]
+  (cond
+    ;; look for filters in the values
+    (map? x) (update-vals x simplify-compound-filter)
```
Comment (metamben) [inline shared/src/metabase/mbql/util.cljc:95]: I'm not convinced there is a real use case when we have to descend into maps, but there is a test for it. @camsaul is this really needed?

File: shared/src/metabase/mbql/util.cljc [hunk 2]
```diff
@@ -64,53 +64,55 @@
...
+                       ;; double negation, eliminate both
+                       :not (recur (second farg))
+                       ;; use de Morgan's law to push the negation down
+                       :and (simplify-and-or-filter :or (map #(vector :not %) (rest farg)))
+                       :or  (simplify-and-or-filter :and (map #(vector :not %) (rest farg)))
+                       x))
+              :and (simplify-and-or-filter :and args)
+              :or  (simplify-and-or-filter :or args)
+              ;; simplify the elements of the vector
+              (mapv simplify-compound-filter x)))))
```
Comment (metamben) [inline shared/src/metabase/mbql/util.cljc:115]: I'm not sure about this either, but again, there is a test case for it.

File: src/metabase/lib/equality.cljc [hunk 3]
```diff
@@ -18,7 +18,9 @@
    [metabase.mbql.util.match :as mbql.u.match]
    [metabase.shared.util.i18n :as i18n]
    [metabase.util.log :as log]
-   [metabase.util.malli :as mu]))
+   [metabase.util.malli :as mu]
+   #?@(:clj  ([clojure.core.memoize :as memoize]))
+   #?@(:cljs ([metabase.util.memoize :as memoize]))))
```
Comment (bshepherdson) [inline src/metabase/lib/equality.cljc:23]: These can and should be combined in a single `#?@(...)` expression:

```suggestion
   #?@(:clj  ([clojure.core.memoize :as memoize])
       :cljs ([metabase.util.memoize :as memoize]))))
```

File: src/metabase/lib/equality.cljc [hunk 4]
```diff
@@ -358,9 +360,9 @@
                  ;; Perhaps any collision should be an error?
                  (merge matches result)))))))
 
-(mu/defn closest-matches-in-metadata :- [:map-of
-                                         lib.metadata/ColumnMetadata
-                                         ::lib.schema.common/int-greater-than-or-equal-to-zero]
+(mu/defn closest-matches-in-metadata* :- [:map-of
```
Comment (bshepherdson) [inline src/metabase/lib/equality.cljc:363]: `^:private`?

File: src/metabase/util.cljc [hunk 5]
```diff
@@ -15,13 +15,15 @@
...
+   #?@(:clj  ([clojure.core.memoize :as memoize]
+              [clojure.math.numeric-tower :as math]
               [metabase.config :as config]
               #_{:clj-kondo/ignore [:discouraged-namespace]}
               [metabase.util.jvm :as u.jvm]
               [metabase.util.string :as u.str]
               [potemkin :as p]
-              [ring.util.codec :as codec]]))
+              [ring.util.codec :as codec]))
+   #?@(:cljs ([metabase.util.memoize :as memoize])))
```
Comment (bshepherdson) [inline src/metabase/util.cljc:26]: Combine these like above.

File: src/metabase/util.cljc [hunk 6]
```diff
@@ -202,22 +204,26 @@
 (def ^{:arglists '([x])} ->kebab-case-en
   "Like [[camel-snake-kebab.core/->kebab-case]], but always uses English for lower-casing, supports keywords with
   namespaces, and returns `nil` when passed `nil` (rather than throwing an exception)."
-  (wrap-csk-conversion-fn-to-handle-nil-and-namespaced-keywords ->kebab-case-en*))
+  (wrap-csk-conversion-fn-to-handle-nil-and-namespaced-keywords
+   (memoize/lru ->kebab-case-en* :lru/threshold 256)))
```
Comment (bshepherdson) [inline src/metabase/util.cljc:208]: We'll only ever pass the same set of perhaps a few hundred values to this - it seems like we could save the LRU overhead and use `clojure.core/memoize` to cache everything forever.

File: src/metabase/lib/equality.cljc [hunk 7]
```diff
@@ -423,6 +425,14 @@
      ;; Prefer the more precise integer ID matches over the name-based ones.
      (merge by-name basic))))
 
+(def ^{:arglists '([needles haystack-metadatas]
+                   [needles haystack-metadatas opts]
+                   [query stage-number needles haystack-metadatas]
+                   [query stage-number needles haystack-metadatas opts])}
+  closest-matches-in-metadata
+  "The cached version of [[closest-matches-in-metadata*]]."
+  (memoize/lru closest-matches-in-metadata* :lru/threshold 8))
```
Comment (metamben) [inline src/metabase/lib/equality.cljc:434]: - To fix #33676, we could cache `breakoutable-columns`, but that spends most of the time in this function and caching this seem more generally useful.
- I expect that eventually the siblings of this function will have to be cached.
- The number 8 is quite random, but I think it should be enough for an action on a page and still not way too much. A lower number would probably work too.
- A TTL cache might be more appropriate here, but I didn't want to spend time porting that too.

File: src/metabase/util.cljc (hunk 6 above)
Comment (metamben) [inline src/metabase/util.cljc:208]: [generic-split](https://github.com/clj-commons/camel-snake-kebab/blob/ac08444c94aca4cba25d86f3b3faf36596809380/src/camel_snake_kebab/internals/string_separator.cljc#L42) in the camel-snake-kebab library turned out to be too expensive.

File: src/metabase/util/memoize.cljs [hunk 8]
```diff
@@ -0,0 +1,130 @@
+(ns metabase.util.memoize
+  "Copied from clojure.core.memoize."
```
Comment (metamben) [inline src/metabase/util/memoize.cljs:2]: This is a simplified port (copy, really) of the parts of clojure.core.memoize needed for implementing LRU caches.

More could be added or we could create a proper CLJS port, but I wanted to get a fix quickly.

File: src/metabase/lib/equality.cljc (hunk 3 above)
Comment (bshepherdson) [reply src/metabase/lib/equality.cljc:23]: Actually on looking further, we should write `src/metabase/util/memoize.clj` that does `potemkin/import-vars` of the handful of public functions from `clojure.core.memoize`. Then we can simply `(:require [metabase.util.memoize :as memoize])` without having to juggle this.

File: src/metabase/util/memoize.cljs [hunk 9]
```diff
@@ -0,0 +1,130 @@
+(ns metabase.util.memoize
```
Comment (bshepherdson) [inline src/metabase/util/memoize.cljs:1]: As noted above, this should have a companion `.clj` file that re-exports a few functions from `clojure.core.memoize`, so the two files have the same API. Then this namespace can be `:require`d naively and will just work on both platforms.

**Review by bshepherdson — CHANGES_REQUESTED** [review]

File: src/metabase/util.cljc (hunk 6 above)
Comment (metamben) [reply src/metabase/util.cljc:208]: I'll need more convincing. (I'm old enough to have lived in a world where Prolog systems had no GC for atoms and felt the pain.)

File: src/metabase/lib/equality.cljc (hunk 4 above)
Comment (metamben) [reply src/metabase/lib/equality.cljc:363]: I did consider that, but thought there could be reasons for wanting to bypass caching.

File: src/metabase/util/memoize.cljs (hunk 9 above)
Comment (metamben) [reply src/metabase/util/memoize.cljs:1]: done

File: src/metabase/lib/equality.cljc (hunk 3 above)
Comment (metamben) [reply src/metabase/lib/equality.cljc]: done

**Review by bshepherdson — APPROVED** [review]

File: src/metabase/lib/equality.cljc (hunk 4 above)
Comment (metamben) [reply src/metabase/lib/equality.cljc:362]: @bshepherdson will deal with this in his PR cleaning up and optimizing these parts of the code.

**Pull request merged**